src/source/Crypto/Dtls_mbedtls.c

```diff
@@ -0,0 +1,739 @@
+
+    CHK(STRCMP(pExpectedFingerprint, actualFingerprint) == 0,  STATUS_SSL_REMOTE_CERTIFICATE_VERIFICATION_FAILED);
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pDtlsSession->sslLock);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS dtlsSessionPopulateKeyingMaterial(PDtlsSession pDtlsSession, PDtlsKeyingMaterial pDtlsKeyingMaterial)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    UINT32 offset = 0;
+    BOOL locked = FALSE;
+    PTlsKeys pKeys;
+    BYTE keyingMaterialBuffer[MAX_SRTP_MASTER_KEY_LEN * 2 + MAX_SRTP_SALT_KEY_LEN * 2];
+    mbedtls_ssl_srtp_profile negotiatedSRTPProfile;
+
+    CHK(pDtlsSession != NULL && pDtlsKeyingMaterial != NULL, STATUS_NULL_ARG);
+    pKeys = &pDtlsSession->tlsKeys;
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+
+    CHK(mbedtls_ssl_tls_prf(pKeys->tlsProfile, 
+                    pKeys->masterSecret,
+                    ARRAY_SIZE(pKeys->masterSecret),
+                    KEYING_EXTRACTOR_LABEL,
+                    pKeys->randBytes,
+                    ARRAY_SIZE(pKeys->randBytes),
+                    keyingMaterialBuffer,
+                    ARRAY_SIZE(keyingMaterialBuffer)) == 0, STATUS_INTERNAL_ERROR);
+    
+    pDtlsKeyingMaterial->key_length = MAX_SRTP_MASTER_KEY_LEN + MAX_SRTP_SALT_KEY_LEN;
+
+    MEMCPY(pDtlsKeyingMaterial->clientWriteKey, &keyingMaterialBuffer[offset], MAX_SRTP_MASTER_KEY_LEN);
+    offset += MAX_SRTP_MASTER_KEY_LEN;
+
+    MEMCPY(pDtlsKeyingMaterial->serverWriteKey, &keyingMaterialBuffer[offset], MAX_SRTP_MASTER_KEY_LEN);
+    offset += MAX_SRTP_MASTER_KEY_LEN;
+
+    MEMCPY(pDtlsKeyingMaterial->clientWriteKey + MAX_SRTP_MASTER_KEY_LEN, &keyingMaterialBuffer[offset], MAX_SRTP_SALT_KEY_LEN);
+    offset += MAX_SRTP_SALT_KEY_LEN;
+
+    MEMCPY(pDtlsKeyingMaterial->serverWriteKey + MAX_SRTP_MASTER_KEY_LEN, &keyingMaterialBuffer[offset], MAX_SRTP_SALT_KEY_LEN);
+
+    negotiatedSRTPProfile = mbedtls_ssl_get_dtls_srtp_protection_profile(&pDtlsSession->sslCtx);
+    switch (negotiatedSRTPProfile) {
+        case MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_80:
+            pDtlsKeyingMaterial->srtpProfile = SRTP_PROFILE_AES128_CM_HMAC_SHA1_80;
+            break;
+        case MBEDTLS_SRTP_AES128_CM_HMAC_SHA1_32:
+            pDtlsKeyingMaterial->srtpProfile = SRTP_PROFILE_AES128_CM_HMAC_SHA1_32;
+            break;
+        default:
+            CHK(FALSE, STATUS_SSL_UNKNOWN_SRTP_PROFILE);
+    }
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pDtlsSession->sslLock);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS dtlsSessionShutdown(PDtlsSession pDtlsSession)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+    INT32 sslRet;
+
+    CHK(pDtlsSession != NULL, STATUS_NULL_ARG);
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
+
+    do sslRet = mbedtls_ssl_close_notify(&pDtlsSession->sslCtx);
```

Comment:
Use curlies even with a single line statements

Reply:
How do we make sure this doesn't evaluate into a tight loop?

Reply:
This is going to tell mbedtls to flush its internal buffer. If a handshake has happened, mbedtls will also send a close notify to the other peer. 

Sending a close notify might fail if sometimes the send callback fails, https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/pull/565/files#diff-1ec7a454de24efa5d640c90c61260220R116-R128. But, since the outgoing callback that's passed to DTLS is always "successful", VOID as return type, mbedtls should be able to always flush its internal buffer, until it runs out or not `MBEDTLS_ERR_SSL_WANT_WRITE`.

Should I add the number of max retries to make sure that this loop will always end?

Reply:
I doubt it... Perhaps, if this proves to be an issue on the MBED TLS side